cmd/maintenance.go

```diff
@@ -74,64 +96,72 @@ func newMaintenanceCMD() *cobra.Command {
 }
 
 func (c *controller) runMaintenance(cmd *cobra.Command, _ []string) error {
+	ctx := cmd.Context()
 
 	kubeClient, err := client.NewWithWatch(ctrl.GetConfigOrDie(), client.Options{
 		Scheme: pkg.SetupScheme(),
 	})
 	if err != nil {
-		return err
+		return fmt.Errorf("failed to initialize kube client: %w", err)
 	}
 
-	var m Maintenance
-	switch serviceName {
-	case postgresql:
+	if err = validateMandatoryEnvs(serviceName); err != nil {
+		return fmt.Errorf("missing required environment variables: %w", err)
+	}
 
-		sgNamespace := viper.GetString("SG_NAMESPACE")
-		if sgNamespace == "" {
-			return fmt.Errorf("missing environment variable: %s", "SG_NAMESPACE")
-		}
-		apiPassword := viper.GetString("API_PASSWORD")
-		if apiPassword == "" {
-			return fmt.Errorf("missing environment variable: %s", "API_PASSWORD")
-		}
+	log := logr.FromContextOrDiscard(ctx).WithValues(
+		"instanceNamespace", viper.GetString("INSTANCE_NAMESPACE"),
+		"claimName", viper.GetString("CLAIM_NAME"),
+		"claimNamespace", viper.GetString("CLAIM_NAMESPACE"),
+		"serviceId", viper.GetString("SERVICE_ID"),
+	)
 
-		apiUserName := viper.GetString("API_USERNAME")
-		if apiUserName == "" {
-			return fmt.Errorf("missing environment variable: %s", "API_USERNAME")
-		}
+	vh := getVersionHandler(kubeClient, log)
 
-		sClient, err := stackgres.New(apiUserName, apiPassword, sgNamespace)
-		if err != nil {
-			return err
-		}
-		m = &maintenance.PostgreSQL{
-			Client:          kubeClient,
-			StackgresClient: sClient,
-			MaintTimeout:    time.Hour,
+	var m Maintenance
+	switch serviceName {
+	case postgresql:
+		if m, err = initPostgreSQL(kubeClient, vh, log); err != nil {
+			return fmt.Errorf("failed to initialize postgresql: %w", err)
 		}
 	case redis:
-		m = maintenance.NewRedis(kubeClient, getHTTPClient())
+		m = maintenance.NewRedis(kubeClient, getHTTPClient(), vh, log)
 
 	case minio:
-		m = maintenance.NewMinio(kubeClient, getHTTPClient())
+		m = maintenance.NewMinio(kubeClient, getHTTPClient(), vh, log)
 
 	case mariadb:
-		m = maintenance.NewMariaDB(kubeClient, getHTTPClient())
+		m = maintenance.NewMariaDB(kubeClient, getHTTPClient(), vh, log)
 
 	case keycloak:
-		m = maintenance.NewKeycloak(kubeClient, getHTTPClient())
+		m = maintenance.NewKeycloak(kubeClient, getHTTPClient(), vh, log)
 
 	case nextcloud:
-		m = maintenance.NewNextcloud(kubeClient, getHTTPClient())
+		m = maintenance.NewNextcloud(kubeClient, getHTTPClient(), vh, log)
 
 	case forgejo:
-		m = maintenance.NewForgejo(kubeClient, getHTTPClient())
+		m = maintenance.NewForgejo(kubeClient, getHTTPClient(), vh, log)
 	default:
 
 		panic("service name is mandatory")
 	}
 
-	return m.DoMaintenance(cmd.Context())
+	if err = errors.Join(
+		m.DoMaintenance(ctx),
```

Comment:
I have decided to continue releasing the latest version even though maintenance might fail

Reply:
Makes sense.

If we ever have a bug in the main maintenance we could never fix it automagically otherwise.

Maintenance is broken -> we release new version that fixes it -> it never gets applied because maintance is broken.